File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpaceRuby/SpecificDamage.rb
require_relative 'Damage'
require_relative 'SpecificDamageToUI'


module Deepspace
    class SpecificDamage < Damage
        
        attr_reader :weapons

        public_class_method def initialize(w,s)
            super(s)
            @weapons = w
        end
        
        def copy()
            return new(@weapons,@nShields)
        end

        def getUIversion
            return SpecificDamageToUI.new(self)
        end 

        def adjust(w, s)
            
            ns_aux = 0

            if @nShields > 0
                if @nShields > s.size
                    ns_aux = s.size
                else
                    ns_aux = @nShields
                end
            end
        

            aux = Array.new(w)

            w_aux = Array.new

            for i in(0..(@weapons.size - 1)) do #recorre el array de armas del damage
                if (arrayContainsType(aux, @weapons.at(i)) != nil) #si dentro de las armas que tengo en posesion esta un arma del tipo i, 
                    w_aux << aux.at(arrayContainsType(aux, @weapons.at(i))).type
                    aux.delete_at(arrayContainsType(aux, @weapons.at(i))) 
                end
            end
            return SpecificDamage.new(w_aux, ns_aux)
            

        end

        def discardWeapon(w)
            if (@weapons != nil)
                for t in @weapons do
                    if t == w.type
                        @weapons.delete(t)
                    end
                end
            end
        end


        def discardShieldBooster
            if @nShields > 0
                @nShields -= 1
            end 
        end

        def hasNoEffect
            return ( (@nShields == 0) && ((@weapons.length == 0 || @weapons == nil)))
        end

        def to_s
            return getUIversion.to_s
        end


        private

        def arrayContainsType(w,t)
            return w.index{ |wea| wea.type == t}
        end

    end
end

File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpace_4/src/deepspace/SpecificDamage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author manbolq
 */
public class SpecificDamage extends Damage{
    private ArrayList<WeaponType> weapons;
    
    /**************************************************************************/
    /**************************    CONSTRUCTOR   ******************************/
    /**************************************************************************/
    
    SpecificDamage(ArrayList<WeaponType> wl, int s){
        super(s);
        weapons = wl;
    }
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/
    
    
    public ArrayList<WeaponType> getWeapons(){
        return weapons;
    }
    
    
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t){
        for (int i = 0; i < w.size(); i++)
            if (w.get(i).getType() == t)
                return i;
        
        return -1;
    }

    @Override
    public Damage copy() {
        return new SpecificDamage(weapons, nShields);
    }

    @Override
    DamageToUI getUIversion() {
        return new SpecificDamageToUI(this);
    }

    @Override
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int ns_aux = 0;
        if (nShields > 0){
            if (nShields > s.size())
                ns_aux = s.size();
            else
                ns_aux = nShields;
        }
        
        ArrayList<Weapon> aux = new ArrayList<>();
        for (int i = 0; i < w.size(); i++)
            aux.add(w.get(i));

        ArrayList<WeaponType> w_aux = new ArrayList<>();

        for (int i = 0; i < weapons.size(); i++){
            if (arrayContainsType(aux, weapons.get(i)) >= 0){
                w_aux.add(aux.get(arrayContainsType(aux, weapons.get(i))).getType());
                aux.remove(arrayContainsType(aux, weapons.get(i)));
            }
        }
        
        return new SpecificDamage(w_aux, ns_aux);
    }

    @Override
    public void discardWeapon(Weapon w) {
        if (weapons != null){
            for (int i = 0; i < weapons.size(); i++){
                if (weapons.get(i) == w.getType())
                    weapons.remove(i);
            }
        }
    }

    @Override
    public boolean hasNoEffect() {
        return ( (nShields == 0) && (weapons == null || weapons.isEmpty()));
    }

    @Override
    public String toString() {
        return getUIversion().toString();
    }
    
}
